### DISCUSSION TITLE
Can I `replace` a screen with `useLinkTo`?

### DISCUSSION THREAD

**Discussion by nandorojo** [react-navigation/react-navigation#10517] [Q&A] [answered]
# Update

Final answer used by Solito here: https://github.com/react-navigation/react-navigation/discussions/10517#discussioncomment-2814967

----

Is it possible to  dispatch a `replace` from `useLinkTo`, while still navigating using a URL only? Something like:

```ts
const linkTo = useLinkTo()

const onPress = () => linkTo('/artists/drake', { replace: true })
```

I know that `linkTo` accepts an action, but I'd like to pass a URL so that my linking config is the source of truth. 

**Maybe the solution is to create a wrapper that reads in the URL and dynamically creates a `replace` action by reading in the linking config/navigation state?**

Thanks so much for your help!

**Comment by satya164** [comment]
It's not that straightforward with nested navigators since depending on the set of navigators that's rendered, you will need to find a navigator that can replace a screen (only stack navigators can).

You can write your own helper that parses the link with `getStateFromPath` and `getActionFromState` and then modify that action how you like it to get a proper replace action.

**Reply by nandorojo** [reply]
@satya164 thanks, really appreciate the fast reply.

I assumed it wouldn't be that simple, so that makes sense.

2 additional questions:

1. Is `LinkingContext` [exposed](https://github.com/react-navigation/react-navigation/blob/main/packages/native/src/index.tsx)? Perhaps I missed it, but I didn't see a way to import it from `@react-navigation/native`.
2. Is there a recommended way to "transform" a given action into a `replace` action? For example, could I check if the `action` is for a stack, and only in that case turn an action with `type: 'NAVIGATE'` into a `replace` action? Would I be able to discern that from `action.payload` somehow?

Presumably I would do something like this, based on your description. Does this seem right?

```js
const navigation = useNavigation()
const linkTo = useLinkTo()
const linking = useContext(LinkingContext) // is this exposed?

const onPress = () => {
  const path = '/artists/drake'

  const { options } = linking

  const state = options?.getStateFromPath
    ? options.getStateFromPath(to, options.config)
    : getStateFromPath(to, options?.config)

  if (state) {
    const action = getActionFromState(state, options?.config)

     // 👋 how should I modify the action here? 
     // pseudocode (how would I do this?)
     if (action.payload.navigator == 'stack') {
       action.type == 'REPLACE'
     }
     // end pseudocode

    if (action !== undefined) {
      navigation.dispatch(action)
    } else {
      navigation.reset(state)
    }
  } else {
    throw new Error('Failed to parse the path to a navigation state.')
  }
}
```

**Reply by nandorojo** [reply]
@satya164 apologies to ping you here, I'm sure you're busy. Just wanted to see if there is a potential solution for turning a navigate action into a replace one after detecting it's in a stack. 

PS - just started sponsoring `react-navigation` 😎

**Comment by kylegwalsh** [comment, marked as answer]
@nandorojo I just ran into this in my own application.

I believe I may have found a solution while running some experiments.

Some background:
- We have a custom hook for useRouter() that we used to extend `react-navigation` methods
- We have a top-level StackNavigator that encompasses our whole application and then several different navigators underneath that

Originally I thought I could extend the getStateFromPath or getActionFromState methods on the linking config directly, but I did not have access to the dispatch method there, so I moved the logic into our userRouter() hook.

In our replace method, I ended up skipping the linkTo(path) logic entirely and just calling a dispatch method with the StackActions.replace() contents.

Here's an example
````tsx
    /** Navigates and removes the current page from the stack */
    replace: useCallback(
      (path) => {
        console.log('[useRouter] Replacing path', path);

        // Web navigation
        if (router) router.replace(path);
        // React native navigation
        else {
          const state = getStateFromPath(path, linking.config);
          // Verify state and then extract action
          if (state) {
            const action: any = getActionFromState(state);
            console.log('Action', action);

            // Dispatch our own replace action
            if (action?.payload?.name) {
              navigation.dispatch(
                StackActions.replace(action?.payload?.name, action?.payload?.params)
              );
            }
            // Fall back to linkTo
            else {
              linkTo(path);
            }
          }
          // Fall back to linkTo
          else {
            linkTo(path);
          }
        }
      },
      [linkTo, router]
    ),
````

Sorry if there are any typos, we have some additional logic to handle query params throughout that method that I removed for clarity.

**Reply by nandorojo** [reply]
Thanks for sharing this! Very helpful. 

I believe this is what I need. Would be nice if someone from @react-navigation could sign off on it being correct. I assume the `action.payload` will always have a `name` and optional `params`. What about things like a nested `screen`, though? I have to check out the types for the `action`.

PS some small suggestions for sharing code samples on GitHub:

Add `tsx` after the first 3 backticks of your code block to add syntax highlighting, like this:

```tsx
```tsx
```

I also suggest not using single-line `else` statements for the example, as it makes it a bit harder to read.

Thanks again for sharing!

**Reply by nandorojo** [reply]
As far as a generalizable approach which I can use in `solito`, I would need `LinkingContext` exposed. I believe I can dangerously import it from react navigation's dist folder for now.

**Reply by kylegwalsh** [reply]
Sure thing. Updated that formatting.

Nested screens seem to be working fine on our end (tested on screens that were pretty deep down).

**Reply by nandorojo** [reply]
Got it, looks like a nested screen would be [forwarded in the `params`](https://github.com/react-navigation/react-navigation/blob/main/packages/core/src/getActionFromState.tsx#L98) so that should work.

**Reply by nandorojo** [reply]
Here's the code I'm using. It has TypeScript support too.

```tsx
replace: useCallback(
      (
        url: Parameters<NextRouterType['replace']>[0],
        as?: Parameters<NextRouterType['replace']>[1],
        options?: TransitionOptions
      ) => {
        if (NextRouter?.router) {
          NextRouter.replace(url, as, options)
        } else {
          const to = parseNextPath(as || url)

          if (to) {
            // custom logic to create a replace() from a URL on native
            // https://github.com/react-navigation/react-navigation/discussions/10517
            const { options } = linking

            const state = options?.getStateFromPath
              ? options.getStateFromPath(to, options.config)
              : getStateFromPath(to, options?.config)

            if (state) {
              const action = getActionFromState(state, options?.config)

              if (action !== undefined) {
                if (
                  'payload' in action &&
                  action.payload &&
                  'name' in action.payload &&
                  action.payload.name
                ) {
                  const { name, params } = action.payload
                  navigation?.dispatch(StackActions.replace(name, params))
                } else {
                  navigation?.dispatch(action)
                }
              } else {
                navigation?.reset(state)
              }
            }
          }
        }
      },
      [linkTo]
    )
```
